File: frontend/src/api/companies.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import { apiClient } from './client'

// ── Types ─────────────────────────────────────────────────────────────────────

export interface AddressSchema {
    street?: string | null
    city?: string | null
    postal_code?: string | null
    country?: string | null
}

export interface CompanyOut {
    id: string
    name: string
    sector?: string | null
    website?: string | null
    address?: AddressSchema | null
    tags: string[]
    notes?: string | null
    contacts_count: number
    created_at: string
    updated_at: string
}

export interface CompanyCreate {
    name: string
    sector?: string | null
    website?: string | null
    address?: AddressSchema | null
    tags?: string[]
    notes?: string | null
}

export type CompanyUpdate = Partial<CompanyCreate>

export interface CompanyList {
    items: CompanyOut[]
    total: number
    page: number
    page_size: number
}

export interface CompanyListParams {
    search?: string
    tag?: string
    page?: number
    page_size?: number
}

// ── API functions ─────────────────────────────────────────────────────────────

export const companiesApi = {
    list: (params?: CompanyListParams) =>
        apiClient.get<CompanyList>('/companies', { params }).then((r) => r.data),

    get: (id: string) =>
        apiClient.get<CompanyOut>(`/companies/${id}`).then((r) => r.data),

    create: (data: CompanyCreate) =>
        apiClient.post<CompanyOut>('/companies', data).then((r) => r.data),

    update: (id: string, data: CompanyUpdate) =>
        apiClient.put<CompanyOut>(`/companies/${id}`, data).then((r) => r.data),

    delete: (id: string) =>
        apiClient.delete(`/companies/${id}`),
}

// ── TanStack Query hooks ──────────────────────────────────────────────────────

export const COMPANIES_KEY = 'companies'

export function useCompanies(params?: CompanyListParams) {
    return useQuery({
        queryKey: [COMPANIES_KEY, params],
        queryFn: () => companiesApi.list(params),
    })
}

export function useCompany(id: string) {
    return useQuery({
        queryKey: [COMPANIES_KEY, id],
        queryFn: () => companiesApi.get(id),
        enabled: !!id,
    })
}

export function useCreateCompany() {
    const qc = useQueryClient()
    return useMutation({
        mutationFn: (data: CompanyCreate) => companiesApi.create(data),
        onSuccess: () => qc.invalidateQueries({ queryKey: [COMPANIES_KEY] }),
    })
}

export function useUpdateCompany() {
    const qc = useQueryClient()
    return useMutation({
        mutationFn: ({ id, data }: { id: string; data: CompanyUpdate }) =>
            companiesApi.update(id, data),
        onSuccess: (_data, { id }) => {
            qc.invalidateQueries({ queryKey: [COMPANIES_KEY] })
            qc.invalidateQueries({ queryKey: [COMPANIES_KEY, id] })
        },
    })
}

export function useDeleteCompany() {
    const qc = useQueryClient()
    return useMutation({
        mutationFn: (id: string) => companiesApi.delete(id),
        onSuccess: () => qc.invalidateQueries({ queryKey: [COMPANIES_KEY] }),
    })
}

File: backend/app/schemas/companies.py
"""Schémas Pydantic — Entreprises."""
import uuid
from typing import Annotated

from pydantic import BaseModel, Field, HttpUrl, model_validator


class AddressSchema(BaseModel):
    street: str | None = None
    city: str | None = None
    zip_code: str | None = None
    country: str | None = Field(default=None, max_length=2)


class CompanyBase(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    vat_number: str | None = Field(default=None, max_length=50)
    siren: str | None = Field(default=None, max_length=20)
    sector: str | None = Field(default=None, max_length=100)
    size: str | None = Field(default=None, max_length=50)
    website: str | None = Field(default=None, max_length=500)
    notes: str | None = Field(default=None, max_length=10000)
    tags: list[str] = Field(default_factory=list)
    addresses: list[AddressSchema] = Field(default_factory=list)


class CompanyCreate(CompanyBase):
    pass


class CompanyUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    vat_number: str | None = Field(default=None, max_length=50)
    siren: str | None = Field(default=None, max_length=20)
    sector: str | None = Field(default=None, max_length=100)
    size: str | None = Field(default=None, max_length=50)
    website: str | None = Field(default=None, max_length=500)
    notes: str | None = Field(default=None, max_length=10000)
    tags: list[str] | None = None
    addresses: list[AddressSchema] | None = None


class CompanyOut(CompanyBase):
    id: uuid.UUID
    contacts_count: int = 0

    model_config = {"from_attributes": True}


class CompanyList(BaseModel):
    items: list[CompanyOut]
    total: int
    page: int
    page_size: int
